### PR TITLE
Allow apps to register an initial state

### LINKED ISSUE #13488: Provide an OCP API for apps to load their initial state
**Is your feature request related to a problem? Please describe.**
Apps usually have some kind of state for the initial rendering of their UI or for a background task. From what I see in our existing code (server and apps), most dynamic code loads its state asynchronously. The static parts (rendered on the server) usually have their state baked into the resulting HTML, hence no data is fetched.

This causes additional requests. Therefore it takes longer for the page to load, the page does not look ready right away, lots of loading animations are shown (best example: the settings pages) and it just gives the application a bit of an instable/fragile appearance.

**Describe the solution you'd like**
Apps (and the server) should try to load data eagerly with the page to be able to render immediately. While there is a bit of an overhead to load all data at once on the server-side, I am convinced that in total there it's less expensive (in terms of computational power) to load all data once than in individual requests.

It's already possible to do that. The usual trick is to fill a hidden input field with a serialized JSON clob. On the client-side, this clob is deserialized into a JS object. An example of this trick can be seen here: https://github.com/nextcloud/server/pull/13481.

This trick above works for apps that load their own template or have a template rendered somewhere else like the settings. However, it does not work for apps that are active on server or other app's pages because there they have no influence on the rendered template.

I would therefore like to propose a new server-side API to store JSON-serializable data that is then merged into a single serialized clob and proved via the client-side OCP.

**Describe alternatives you've considered**
I don't have any at the moment except for not improving this.

**Additional context**
This helps with the "we have way too many requests on page load" issue.

cc @nextcloud/javascript let me know what you think. I'd work on this ASAP so that we can use this in Nextcloud 16 already and have pages load faster :rocket:

### PR DESCRIPTION
**Pull request by rullzer**
Fixes #13488 

This introduces a way for apps to register their initial state so this does not have to be fetched async.

Todo:
- [x] Clean OCP javascript API to obtain your data
- [ ] Tests for the service

### CODE REVIEW

File: core/templates/layout.user.php [hunk 1]
```diff
@@ -28,6 +28,10 @@
 	<body id="<?php p($_['bodyid']);?>">
 	<?php include 'layout.noscript.warning.php'; ?>
 
+		<?php foreach ($_['initialStates'] as $app => $initialState) { ?>
+			<input type="hidden" id="initial-state-<?php p($app); ?>" value="<?php p(base64_encode($initialState)); ?>">
+		<?php }?>
```
Comment (MorrisJobke) [inline core/templates/layout.user.php]: Limitation of this is that it only works for logged in pages. I would just add it to every page if there is some initial state.

File: apps/twofactor_backupcodes/src/settings.js [hunk 2]
```diff
@@ -4,7 +4,7 @@ import store from './store';
 
 Vue.prototype.t = t;
 
-const initialStateElem = document.getElementById('twofactor-backupcodes-initial-state');
+const initialStateElem = document.getElementById('initial-state-twofactor_backupcodes');
```
Comment (ChristophWurst) [inline apps/twofactor_backupcodes/src/settings.js]: I guess you want to change that to the fancy new API ;)

File: core/src/OCP/initialstate.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+/**
+ * @namespace OCP
+ * @class InitialState
+ */
+export default {
+	loadState: function(app) {
+		const elem = document.querySelector('#initial-state-' + app);
+		if (elem === null) {
+			return {};
```
Comment (ChristophWurst) [inline core/src/OCP/initialstate.js]: I would prefer to have a `console.warn` here

File: core/src/OCP/initialstate.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+		}
+
+		try {
+			return JSON.parse(atob(elem.value));
+		} catch (e) {
+			console.log('Could not parse initial state of ' + app);
```
Comment (ChristophWurst) [inline core/src/OCP/initialstate.js]: make that a `console.error` and `throw new Error('...')` afterwards, then apps can react to it and do a fallback

**Review by ChristophWurst — CHANGES_REQUESTED** [review]

**Comment by rullzer** [comment]
@MorrisJobke @ChristophWurst comments address. Will rebase and squash with the green light ;)

**Review by ChristophWurst — APPROVED** [review]
Looks great!

**Review by MorrisJobke — APPROVED** [review]
Beside my comment this looks good 👍

File: lib/private/InitialStateService.php [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+	}
+
+	public function provideInitialState(string $appName, $data) {
+		// Scalars and JsonSerializable are fine
+		if (is_scalar($data) || $data instanceof \JsonSerializable || is_array($data)) {
+			$this->states[$appName] = json_encode($data);
+			return;
+		}
+
+		$this->logger->warning('Invalid data provided to provideInitialState');
```
[the remark this answers is not in the document]
Comment (rullzer) [reply lib/private/InitialStateService.php]: Done

**Review by skjnldsv — COMMENTED** [review]
You'll add tests later? (looking at the OP) :wink:

**Pull request merged**